_posts/2020-01-29-#17487.md

```diff
@@ -0,0 +1,115 @@
+
+Because the UTXO set is accessed so frequently, we would like its contents
+available as quickly as possible. This amounts to having as much of the set
+in-memory as possible. The issue is that (at the moment) the in-memory
+representation of the UTXO set is more than 8GB, and obviously not all hosts
+running Bitcoin have that much memory. 
+
+For that reason, the UTXO cache [is stratified across several
+layers](https://jameso.be/dev++2018/#54): some on-disk, and some in-memory. The
+`-dbcache` parameter controls how much memory we allocate to the in-memory
+portion. As we validate blocks, we pull unspent coins that we look up from disk
+into memory until we run out of allocated memory (as indicated by [this
+logic](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L2198-L2220)).
+
+At that point, we completely empty the UTXO cache by writing it to disk by
+calling
+[`CCoinsViewCache::Flush()`](https://github.com/bitcoin/bitcoin/blob/master/src/coins.cpp#L205-L210).
+In master, this is the only way of reconciling the state of the in-memory cache
+with the leveldb store on disk, even though sometimes we Flush() not because we
+have exceeded our dbcache, but to ensure durability. For example, we
+[periodically
+flush](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L2275-L2276)
+the coins cache to avoid having to replay blocks if we shutdown improperly.
+
+Once we flush the in-memory cache, we are forced to read from and write to disk
+for all UTXO operations, which can be notably slower depending on the
+underlying disk. For this reason, separating the emptying of the cache
+from the writing to disk might allow us to ensure durability without losing the
+performance benefits of maintaining the cache.
+
+A year ago, @andrewtoth
+[proposed](https://github.com/bitcoin/bitcoin/pull/15218) that we flush the
+UTXO set after the completion of initial block download to avoid having to
+reconstruct the entire set if an unclean shutdown happened before a periodic
+flush, which basically amounts to a `-reindex-chainstate`. Other reviewers
+criticized this idea because of the performance implications of emptying the
+cache.
+
+Another case that requires writing to disk without necessarily emptying the
+cache can be found in the [assumeutxo
+project](https://github.com/jamesob/assumeutxo-docs/tree/master/proposal). When
+loading a UTXO set from a serialized snapshot, it's preferable to write out the
+newly constructed chainstate immediately after load to avoid having to reload
+the snapshot once again after a bad shutdown. [Benchmarks have
+shown](https://github.com/bitcoin/bitcoin/pull/17487#issuecomment-561741590)
+that some platforms benefit significantly from maintaining the contents of the
+cache after writing them to disk.
+
+### Consensus bug
+
+One interesting dimension of this PR is that in a previous version of it, I had
+actually proposed a consensus bug. I wish I could pass this off as a test of my
```

Comment:
No need for the soul-searching here. I think you can drop the second sentence entirely.

Perhaps you can add some more information about the consensus bug?

Edit: I see that you ask the question and link to the relevant comment below. I think you can just remove this section.